File: simulator/demo/src/main/java/com/example/FraudDetectorService.java
package com.example;

import org.apache.kafka.clients.consumer.ConsumerRecord;

public class FraudDetectorService {
  private final String topic = "ECOMMERCE_NEW_ORDER";

  public static void main(String[] args) {
    var fraudDetectorService = new FraudDetectorService();
    try (var service = new KafkaService(FraudDetectorService.class.getSimpleName(), fraudDetectorService.topic,
        fraudDetectorService::parser)) {
      service.run();
    }
  }

  private void parser(ConsumerRecord<String, String> record) {
    System.out.println("-----------------------------------------");
    System.out.println("Processing new order, checking for fraud");
    System.out.println("Key " + record.key());
    System.out.println("Value " + record.value());
    System.out.println("Partition " + record.partition());
    System.out.println("Offset " + record.offset());

    try {
      Thread.sleep(500);
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
    System.out.println("Order proccessed");
  }
}

File: simulator/project_python/fraud_detector_service.py
from asyncore import dispatcher
import json
from kafka_service import KafkaService
from kafka_dispatcher import KafkaDispatcher


class FraudDetectorService:
    topics = "ECOMMERCE_NEW_ORDER"
    dispatcher = KafkaDispatcher()

    def __init__(self):
        service = KafkaService(FraudDetectorService.__name__, self.topics, self.parser)
        try:
            service.run()
        except Exception as e:
            service.close(e)

    def parser(self, record):
        print("-----------------------------------------")
        print("Processing new order, checking for fraud")
        print("Key => {}".format(record.key))
        print("Value => {}".format(record.value))
        print("Partition => {}".format(record.partition))
        print("Offset => {}".format(record.offset))
        print("Order proccessed\n")

        order = json.loads(record.value)

        if self.getFraud(order):
            dispatcher.send(
                "ECOMMERCE_ORDER_REJECT", key=record.key, value=record.value
            )
            print("FRAUD DETECTED => {}".format(order.get("amount")))
        else:
            dispatcher.send(
                "ECOMMERCE_ORDER_APPROVED", key=record.key, value=record.value
            )
            print("APROVED ORDER  USER ID => {}".format(record.key))

    def getFraud(self, order):
        return order.get("amount") >= 4500


FraudDetectorService()
